### ISSUE TITLE
Authentication failure in Travis push_tag script

### ISSUE THREAD

**Issue by ssoloff** [triplea-game/triplea#1586]
I noticed this when I created a branch in my fork to run the `deploy` phase on Travis.  It looks like there's an authentication failure when the `push_tag` script attempts to delete the remote tag (from build [3901](https://travis-ci.org/triplea-game/triplea/builds/214895446)):

```
$ chmod +x ./.travis/push_tag && ./.travis/push_tag
Configure Git and PUSH A TAG - this trigger and execute the deploy section of this same build file, in the same travis linux container (skip_cleanup setting below is important)
Delete previous tag if it exists, this way we can restart builds and regenerate artifacts in case anything goes wrong
error: tag '1.9.0.0.3901' not found.
remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/triplea-game/triplea.git/'
Push a tag to trigger github releases deployment
```

In this case, the error is not fatal because the remote tag does not already exist (i.e. this is not a rebuild).  Thus, there won't be a duplicate tag failure when the tag is pushed a few lines later.

I suspect the problem is due to using the `origin` repo on line 9:

```
git push origin ":refs/tags/$TAGGED_VERSION"
```

which is effectively read-only because the repo URL specified when cloning does not include authentication credentials in the authority (from line 112 of the Travis log):

```
$ git clone --depth=50 --branch=master https://github.com/triplea-game/triplea.git triplea-game/triplea
```

Should the repo URL be explicitly specified with the GitHub personal access token?  Similar to what is done when the new tag is pushed on line 13:

```
git push -q https://$GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS@github.com/triplea-game/triplea --tags 2>&1 | sed 's|https://.*github|https://[secure]@github|'
```

**Comment by RoiEXLab** [comment]
As you may have noticed, new "github pre-release" files are published...
Because you haven't setup any environment vars inside travis, vars like `$GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS@` are empty strings.
Thats why the script isn't working...

 I can't think of a way to fix this problem for specific forks right now...

**Comment by ssoloff** [comment]
@RoiEXLab Thanks for looking at this.  It sounds like I didn't state the problem clearly in my description.  The snippet of Travis error log I pasted above is from build 3901 of `triplea-game/triplea:master`, **not** from my fork.  Per the same log linked above, you can see that `GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS` is defined on line 151.

(But your point is well taken, and I did indeed run into the problem you described on my own fork.  Obviously, I had to create my own GitHub personal access token, define it in my Travis environment, and then tweak `push_tag` so that it pushed to my fork rather than the official TripleA repo.  That's how I was able to provide [releases](https://github.com/ssoloff/triplea/releases) for others to test #1584).

**Comment by RoiEXLab** [comment]
Ahh I see...
Yes, I believe using an auth token for those steps too is the right way to go!

**Comment by DanVanAtta** [comment]
@ssoloff  (sry for not checking) but for context have you observed this is as happening all of the time, or just one-off failures?

**Comment by ssoloff** [comment]
@DanVanAtta I believe it happens on each build of `triplea-game/triplea:master`.  Here is the relevant portion of the log from the latest `master` [build](https://travis-ci.org/triplea-game/triplea/builds/220366201) (1.9.0.0.4017):

```
$ chmod +x ./.travis/push_tag && ./.travis/push_tag
Configure Git and PUSH A TAG - this trigger and execute the deploy section of this same build file, in the same travis linux container (skip_cleanup setting below is important)
Delete previous tag if it exists, this way we can restart builds and regenerate artifacts in case anything goes wrong
error: tag '1.9.0.0.4017' not found.
remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/triplea-game/triplea.git/'
Push a tag to trigger github releases deployment
```

**Comment by ssoloff** [comment]
I just noticed that the final push in this script is now failing, as well.  Build [4189](https://travis-ci.org/triplea-game/triplea/builds/229826584) was the last build to not exhibit the new failure mode; build [4196](https://travis-ci.org/triplea-game/triplea/builds/230141924) was the first build to exhibit the new failure mode.

From the build 4196 log:

```
$ chmod +x ./.travis/push_tag && ./.travis/push_tag
Configure Git and PUSH A TAG - this trigger and execute the deploy section of this same build file, in the same travis linux container (skip_cleanup setting below is important)
Delete previous tag if it exists, this way we can restart builds and regenerate artifacts in case anything goes wrong
error: tag '1.9.0.0.4196' not found.
remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/[secure]-game/[secure].git/'
Push a tag to trigger github releases deployment
remote: Invalid username or password.
fatal: Authentication failed for 'https://[secure]@github.com/[secure]-game/[secure]/'
```

The last two lines are not present in build 4189.

Note the `[secure]` strings in the log output.  I believe this new failure mode is related to the recently-reported [Travis security advisory](https://blog.travis-ci.com/2017-05-08-security-advisory).  I also noticed the `update_checkstyle_thresholds` script is now failing due to an authentication failure, as well (I'll write up a separate issue for that).

I'm wondering if the triplea-game org's OAuth token was disclosed in an archived Travis log at some time in the past, and as a result of the security advisory, GitHub has revoked that token.

@ron-murhammer, @DanVanAtta: Could one of you please check to see if the GitHub token used by Travis has been revoked?

**Comment by ssoloff** [comment]
One other thing I observed is that new tags are still being created even though the `push_tag` script is failing.  I suspect the Travis GitHub releases provider is creating the tag at the same time it deploys the release.  So the `push_tag` script may be superfluous and probably can be removed.

**Comment by ssoloff** [comment]
The second part of `push_tag` is working again (see [here](https://github.com/triplea-game/triplea/issues/1700#issuecomment-306362347)).  The first part continues to fail for the reason listed at the top of this issue.

@ron-murhammer, @DanVanAtta: `push_tag` wasn't working for almost a month and yet the tags appear to have been created correctly during this time (at least they appear correct to me; someone else please confirm if this is not the case).  I suspect the Travis GitHub releases provider is creating the tag at the same time it deploys the release.  Do you think it's appropriate to remove this script from the build since it apparently offers no additional value?

**Comment by DanVanAtta** [comment]
Digging into this a bit, the tag that is pushed to github is very specific since it is to our format, it is not something travis would do automatically without some configuration.

Chasing this down, note this output line in the travis script: https://github.com/triplea-game/triplea/blob/master/.travis/push_tag#L11

In the failure output, note all error messages are above that line. So what was happening we were failing to delete an existing tag, then following through to the next lines and were successfully pushing a new tag.

This can be seen here:
![push_tag_output](https://cloud.githubusercontent.com/assets/12397753/26813200/782dd00c-4a31-11e7-88cc-4540abc48c40.png)

The delete tag logic was there to help the case where a bad build was pushed by travis and we wanted to regenerate the installers. With the delete tag, you can just restart the build and it'll recreate the release. Trying this out, I see that it is broken:
![rebuild](https://cloud.githubusercontent.com/assets/12397753/26813202/7c97fdf2-4a31-11e7-9d5c-49371a5c98d4.png)

That then lead me to find the first thing, that we were failing to delete tags. This PR should fix up the delete tag task: #1823, and also adds some documentation at the top of the script to better capture the scripts purpose.

**Comment by ssoloff** [comment]
@DanVanAtta I'm a bit confused.

Presumably, the second part of `push_tag` (`Push a tag to trigger github releases deployment`) started working again when you fixed the `GITHUB_PERSONAL_ACCESS_TOKEN_FOR_TRAVIS` issue within the last 24 hours.  Note that in build [4547](https://travis-ci.org/triplea-game/triplea/builds/239463759) (from ~24 hours ago), **both** the delete tag and push tag logic were failing:

```
$ ./.travis/push_tag
Configure Git and PUSH A TAG - this trigger and execute the deploy section of this same build file, in the same travis linux container (skip_cleanup setting below is important)
Delete previous tag if it exists, this way we can restart builds and regenerate artifacts in case anything goes wrong
error: tag '1.9.0.0.4547' not found.
remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/[secure]-game/[secure].git/'
Push a tag to trigger github releases deployment
remote: Invalid username or password.
fatal: Authentication failed for 'https://[secure]@github.com/[secure]-game/[secure]/'
```

Yet tag `1.9.0.0.4547` exists in the repo:

```
$ git tag | grep 4547
1.9.0.0.4547
```

So, if `push_tag` failed to push, but the tag exists, who was responsible for creating it?

If you look at the above log starting on line 5144 (`Preparing deploy`), there appears to be some kind of repo interaction going on that ends with `Current tag is: 1.9.0.0.4547`.  Obviously, `push_tag` successfully creates the tag in the **local** repo on Travis, but I'm wondering if as part of the GitHub release, Travis is doing something equivalent to a `git push --tags`, and that inadvertently (but beneficially) pushes the tag `push_tag` created but failed to push itself?

**Comment by DanVanAtta** [comment]
@ssoloff  that could be right. Do we have a major time gap in tags? Did it ever actually really break?

**Comment by DanVanAtta** [comment]
There does not appear to be a gap in tags.

If we want to determine this for sure, we can do a one-line experiment removing the explicit tag push.  (independently - I think there is a question of if we would want to be dependent on a silent tag push by travis)

**Comment by ssoloff** [comment]
@DanVanAtta I think I know what's going on.

If you look at the code for the Travis GitHub releases provider, you'll see that it [reads the tag](https://github.com/travis-ci/dpl/blob/master/lib/dpl/provider/releases.rb#L23) from the local repo (if `TRAVIS_TAG` is not present, which I don't believe it is in our case).  It then passes this in the [call](https://github.com/travis-ci/dpl/blob/master/lib/dpl/provider/releases.rb#L104) to the GitHub API endpoint that creates a release.

After further consideration, I no longer believe it is Travis doing a silent `git push --tags`.  According to the GitHub API [docs for creating releases](https://developer.github.com/v3/repos/releases/#create-a-release), the tag name is required.  I suspect that if a tag with the specified name does not yet exist when creating the release, GitHub creates the tag.

I tested this by manually creating a release in my fork.  If I specify a tag that doesn't exist, the page indicates it will be created:

![release-new-tag](https://cloud.githubusercontent.com/assets/4826349/26814271/4ea79030-4a52-11e7-990b-5709f5fef6c7.png)

However, if I specify an existing tag, that message is not displayed:

![release-existing-tag](https://cloud.githubusercontent.com/assets/4826349/26814272/53f950fa-4a52-11e7-8b58-cb70794cfb7a.png)

In summary, I think if `push_tag` simply creates the tag in the local repo, it will be created on the remote when the release is deployed.  However, I'm not sure what will happen when triggering a rebuild.  If necessary, I can run an experiment in my fork to test what will happen when rebuilding.

**Comment by ssoloff** [comment]
Closing this issue as both commands in `push_tag` now run without an authentication error at least as of build [4768](https://travis-ci.org/triplea-game/triplea/builds/241761292#L4874):

```
$ ./.travis/push_tag $TAGGED_VERSION
Attempting to delete tag 1.9.0.0.4768, this is so we will recreate the release if this build is being re-run. Ignore any errors from this
remote: warning: Deleting a non-existent ref.        
Done deleting existing tag if it existed.
Pushing tag: 1.9.0.0.4768 to github to trigger releases deployment
```

**Comment by ssoloff** [comment]
Based on what I wrote in a previous [comment](https://github.com/triplea-game/triplea/issues/1586#issuecomment-306381911), I attempted to simplify the Travis build with respect to tag creation.  I thought it was going to be as easy as:

* Removing the `push_tag` script.
* Adding `export TRAVIS_TAG=$TAGGED_VERSION` to `.travis.yml`.

However, it quickly dawned on me that it was a total hack because we should never modify a `TRAVIS_*` environment variable.

After some further research, I got the impression that we're fighting some Travis conventions in how we create releases.  I came up with an alternative solution to create tags and deploy releases in a manner more consistent with Travis conventions.  However, I'm still not convinced it's the right way to go, and I don't want to spend any more time on it.  Therefore, I'm attaching an experimental [patch](https://github.com/triplea-game/triplea/files/1070221/triplea-change-how-tags-are-created-from-travis.patch.txt) to this issue that demonstrates the solution for future reference.

Highlights of the patch:

* A GitHub release is only created from a tag build, never from a branch build.
* Tags are created using the GitHub API rather than via the Git client.
* Release artifacts are properly overwritten if a tag build is restarted.

However, there is one caveat to this solution that may make it incompatible with the current TripleA release process.  Because a GitHub release is now only created from a tag build, the release artifact names should no longer be based on the Travis build number.  If they were, then the names would differ from the tag.  This is best demonstrated with some examples (assume the TripleA engine version is 1.9.0.0 and the most recent Travis build number was 99 in what follows).

First, assume we use the build number when naming build artifacts (what is done today):

* Build number 100 on branch `master` completes successfully and creates tag `1.9.0.0.100`.
* Creation of tag triggers build number 101.
* Build number 101 completes successfully with artifact names containing the identifier `1.9.0.0.101`.
* Build artifacts are used to create GitHub release `1.9.0.0.100`.

Thus, release `1.9.0.0.100` will contain files identified as `1.9.0.0.101`.  This could be somewhat confusing as the originating tag of the release has been lost from the file names.

Second, assume we use the tag when naming build artifacts (what the attached patch does):

* Build number 100 on branch `master` completes successfully and creates tag `1.9.0.0.100`.
* Creation of tag triggers build number 101.
* Build number 101 completes successfully with artifact names containing the identifier `1.9.0.0.100`.
* Build artifacts are used to create GitHub release `1.9.0.0.100`.

Now, release `1.9.0.0.100` contains files identified as `1.9.0.0.100`, so there is no confusion about the originating tag.  However, we have lost the true build number that produced the artifacts.

So both solutions have some deficiency that removes some aspect of identifying information from the build artifact names.  A third possibility would be to name tags without reference to build number, but something that is constant regardless of how many times it's built, such as the commit ref associated with the tag.  Then during the tag build, both the tag name and the build number could be embedded into the artifact names to ensure both tag and originating build are traceable.  For example, assuming the HEAD commit ref on `master` is `abc7def`:

* Build number 100 on branch `master` completes successfully and creates tag `1.9.0.0.abc7def`.
* Creation of tag triggers build number 101.
* Build number 101 completes successfully with artifact names containing the identifier `1.9.0.0.abc7def-101` (i.e. `$TRAVIS_TAG-$TRAVIS_BUILD_NUMBER`).
* Build artifacts are used to create GitHub release `1.9.0.0.abc7def`.

However, usage of the commit ref in the tag name and build artifact names may be incompatible with the future versioning direction of the  project (c.f. #1739).
